cpp/bench/ann/src/diskann/diskann_wrapper.h

```diff
@@ -0,0 +1,335 @@
+    uint32_t num_threads        = omp_get_num_procs() / 2;
+    uint32_t num_nodes_to_cache = 10000;
+    int beam_width              = 2;
+    // Mode metric_objective;
+  };
+
+  diskann_ssd(Metric metric, int dim, const build_param& param);
+
+  void build(const T* dataset, size_t nrow) override;
+
+  void set_search_param(const search_param_base& param) override;
+
+  void search(const T* queries,
+              int batch_size,
+              int k,
+              algo_base::index_type* neighbors,
+              float* distances) const override;
+
+  void save(const std::string& path_to_index) const override;
+  void load(const std::string& path_to_index) override;
+  diskann_ssd(const diskann_ssd<T>& other) = default;
+  std::unique_ptr<algo<T>> copy() override { return std::make_unique<diskann_ssd<T>>(*this); }
+
+  [[nodiscard]] auto get_preference() const -> algo_property override
+  {
+    algo_property property;
+    property.dataset_memory_type = MemoryType::kHost;
+    property.query_memory_type   = MemoryType::kHost;
+    return property;
+  }
+
+ private:
+  std::string index_build_params_str;
+  std::shared_ptr<diskann::PQFlashIndex<T, uint32_t>> p_flash_index_;
+  int beam_width_;
+  uint32_t num_nodes_to_cache_;
+
+  // in-memory index params
+  uint32_t build_pq_bytes_ = 0;
+  uint32_t max_points_;
+  // for safe scratch space allocs, set the default to half the number of procs for loading the
+  // index. User must ensure that the number of search threads is less than or equal to this value
+  int num_search_threads_ = omp_get_num_procs() / 2;
+  // L_search is hardcoded to the maximum visited list size in the search params. This default is
+  // for loading the index
+  uint32_t L_search_ = 384;
+  Mode bench_mode_;
+  std::string base_file_;
+  std::string index_path_prefix_;
+  std::shared_ptr<AlignedFileReader> reader = nullptr;
+};
+
+template <typename T>
+diskann_ssd<T>::diskann_ssd(Metric metric, int dim, const build_param& param) : algo<T>(metric, dim)
+{
+  // Currently set the indexing RAM budget and the search RAM budget to max value to avoid sharding
+  uint32_t build_dram_budget  = std::numeric_limits<uint32_t>::max();
+  uint32_t search_dram_budget = std::numeric_limits<uint32_t>::max();
+  index_build_params_str =
+    std::string(std::to_string(param.R)) + " " + std::string(std::to_string(param.L_build)) + " " +
+    std::string(std::to_string(search_dram_budget)) + " " +
+    std::string(std::to_string(build_dram_budget)) + " " +
+    std::string(std::to_string(param.num_threads)) + " " + std::string(std::to_string(false)) +
+    " " + std::string(std::to_string(false)) + " " + std::string(std::to_string(0)) + " " +
+    std::string(std::to_string(param.QD));
+  base_file_         = param.dataset_file;
+  index_path_prefix_ = param.path_to_index;
+}
+
+template <typename T>
+void diskann_ssd<T>::build(const T* dataset, size_t nrow)
+{
+  diskann::build_disk_index<float>(base_file_.c_str(),
```

Comment:
Where is this defined? Is it using the CPU DiskANN repo to build the ssd verson of the index?

Reply:
Yes, that is right. It is just to benchmark the CPU build